packages/typegpu/src/data/vertexAttribute.ts

```diff
@@ -0,0 +1,458 @@
+          throw new Error('Invalid underlying type');
+      }
+    }
+
+    const vector = this.underlyingType as
+      | Vec2u
+      | Vec3u
+      | Vec4u
+      | Vec2f
+      | Vec3f
+      | Vec4f
+      | Vec2i
+      | Vec3i
+      | Vec4i;
+    const primitive =
+      vectorKindToPrimitive[vector.label as keyof typeof vectorKindToPrimitive];
+
+    const values = new Array(this.elementCount);
+    const reader = new BufferReader(readBuffer);
+    for (let i = 0; i < this.elementCount; i++) {
+      values[i] = readSizedPrimitive(
+        primitive,
+        this.elementSize,
+        reader,
+        this.isSigned,
+      );
+    }
+
+    switch (vector) {
+      case vec2u:
+      case vec2f:
+      case vec2i:
+        return vector(values[0], values[1]) as Parsed<
+          typeof this.underlyingType
+        >;
+      case vec3u:
+      case vec3f:
+      case vec3i:
+        return vector(values[0], values[1], values[2]) as Parsed<
+          typeof this.underlyingType
+        >;
+      case vec4u:
+      case vec4f:
+      case vec4i:
+        return vector(values[0], values[1], values[2], values[3]) as Parsed<
+          typeof this.underlyingType
+        >;
+      default:
+        throw new Error('Invalid underlying type');
+    }
+  }
+
+  measure(
+    _: Parsed<FormatToWGSLType<T>, Record<string, never>> | MaxValue,
+    measurer: IMeasurer = new Measurer(),
+  ): IMeasurer {
+    return measurer.add(this.size);
+  }
+}
+
+const vectorKindToPrimitive = {
+  vec2u: 'u32',
+  vec3u: 'u32',
+  vec4u: 'u32',
+  vec2f: 'f32',
+  vec3f: 'f32',
+  vec4f: 'f32',
+  vec2i: 'i32',
+  vec3i: 'i32',
+  vec4i: 'i32',
+} as const;
+
+function writeSizedVector(
+  output: ISerialOutput,
+  value: vec2u | vec3u | vec4u | vec2f | vec3f | vec4f | vec2i | vec3i | vec4i,
+  elementSize: 1 | 2 | 4,
+  isSigned: boolean,
+): void {
+  const primitive = vectorKindToPrimitive[value.kind];
+  if (!primitive) throw new Error('Invalid vector kind');
+  for (const entry of value) {
+    writeSizedPrimitive(primitive, entry, elementSize, output, isSigned);
+  }
+}
+
+function writeSizedPrimitive(
+  primitive: 'u32' | 'f32' | 'i32',
+  value: number,
+  size: 1 | 2 | 4,
+  output: ISerialOutput,
+  floatSigned = true,
+): void {
+  const buffer = new ArrayBuffer(size);
+  const view = new DataView(buffer);
+
+  const writef8 = (offset: number, value: number) => {
+    const asInt = floatSigned ? value * 127 + 128 : value * 255;
+    view.setUint8(offset, Math.floor(asInt));
+  };
+  const writef16 = (offset: number, value: number) => {
+    const asInt = floatSigned ? value * 32767 + 32768 : value * 65535;
+    view.setUint16(offset, Math.floor(asInt));
+  };
+
+  const setters = {
+    u32: [view.setUint8, view.setUint16, view.setUint32],
+    f32: [writef8, writef16, view.setFloat32],
+    i32: [view.setInt8, view.setInt16, view.setInt32],
+  };
+
+  const setter = setters[primitive][Math.log2(size)];
+  if (setter) setter.call(view, 0, value);
```

Comment:
use block with curly braces to match our convention (also in a few other places in this file)